### PR TITLE
add new metaDuration flag

### PR DESCRIPTION
**Pull request by aojea**
A common problem in golang is to differentiate between default and empty values. This causes a lot of headeaches when defining APIs with golang types, that is even more complex if we add the necessity to define this values via a configuration file or a flag.

It is common to use the metav1.Duration type to define durations, this patch adds a new flag type that allows to define a flag that is able to use a pointer to a metav1.Duration variable.

/kind feature
```release-note
NONE
```

This could have helped in https://github.com/kubernetes/kubernetes/issues/119325, and can help in https://github.com/kubernetes/kubernetes/pull/120354 to be able to use  pointers to the variables with type `metav1.Duration` and avoid the complexity of dealing with golang default values

### CODE REVIEW

File: staging/src/k8s.io/component-base/cli/flag/meta_duration.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+var _ goflag.Value = &MetaDuration{}
+var _ pflag.Value = &MetaDuration{}
+
+// NewMetaDuration takes a pointer to a metav1.Duration and returns the
+// MetaDuration flag parsing shim for that variable
+func NewMetaDuration(m *metav1.Duration) MetaDuration {
+	return MetaDuration{m: m}
+}
+
+// Set implements github.com/spf13/pflag.String
```
Comment (pohly) [inline staging/src/k8s.io/component-base/cli/flag/meta_duration.go:42]:
```suggestion
// String implements github.com/spf13/pflag.String
```

File: staging/src/k8s.io/component-base/cli/flag/meta_duration.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	if err != nil {
+		return err
+	}
+	m.m = &metav1.Duration{Duration: duration}
+	return nil
+}
+
+// Type implements github.com/spf13/pflag.Value
+func (m *MetaDuration) Type() string {
+	return "metav1.Duration"
```
Comment (pohly) [inline staging/src/k8s.io/component-base/cli/flag/meta_duration.go:65]: This will be used visible in `--help`. Perhaps just "duration"?

**Review by pohly — COMMENTED** [review]
> A common problem in golang is to differentiate between default and empty values.

I don't see how this PR helps with that. Was that just to explain why `metav1.Duration` is being used?

**Comment by aroradaman** [comment]
> I don't see how this PR helps with that. Was that just to explain why `metav1.Duration` is being used?

If we use `metav1.Duration`  as a type for a variable in the config we can't differentiate between the default and empty value, and `*metav1.Duration` can't be directly consumed in `pflag.DurationVar` as it expects an object of type `time.Duration` and if the config variable is nil/uninitialized accessing `Duration` field from that variable will lead to panic/sigsegv.

This PR adds a custom flag `cliflag.MetaDuration` which consumes `*metav1.Duration` allowing us to differentiate between default and empty values.

Related thread: https://github.com/kubernetes/kubernetes/pull/120354#discussion_r1316379058

**Comment by pohly** [comment]
> This PR adds a custom flag cliflag.MetaDuration which consumes *metav1.Duration allowing us to differentiate between default and empty values.

Something doesn't add up here for me. To instantiate a `cliflag.MetaDuration`, you always have to have a `metav1.Duration`. How do you then distinguish between "set" and "not set"?

Can you perhaps show an example of how this will be used?

**Comment by aroradaman** [comment]
> Something doesn't add up here for me. To instantiate a cliflag.MetaDuration, you always have to have a metav1.Duration. How do you then distinguish between "set" and "not set"?

This **MetaDuration** flag is instantiated with `*metav1.Duration` not `metav1.Duration`.
Since this is a pointer we can easily distinguish between "set" and "not set"


Example:

```go
// dummy config
type configv1 struct{
    UDPStreamTimeout *metav1.Duration
}

// load config with defaults
config := NewConfig()


// command line flag (syncing config.UDPStreamTimeout with command line flag)
// config.UDPStreamTimeout is a pointer not a value
fs.Var(&NewMetaDuration(config.UDPStreamTimeout), ...)

// check for not set
if config.UDPStreamTimeout != nil{
}
```

Following is the flow without existence of `cliflag.MetaDuration`
```go
fs.DurationVar(config.UDPStreamTimeout.Duration)

// if config.UDPStreamTimeout is "not set" then this will lead to panic/sigsegv 
// and hence we currently cant use *metav1.Duration as type for UDPStreamTimeout
```

I think we need to update `func (m *MetaDuration) Set(val string) error` to not throw an error if m.m is nil, should simply override the value for m.m syncing the original variable `config.UDPStreamTimeout` giving precedence to command line flag over config.

Using `cliflags.MetaDuration` will ultimately allow us to use `*metav1.Duration` instead of `metav1.Duration` as type for `config.UDPStreamTimeout` which currently is not possible.

**Comment by pohly** [comment]
I don't think your example works as intended:

```
config := NewConfig() // UDPStreamTimeout is nil, right?
fs.Var(&NewMetaDuration(config.UDPStreamTimeout /* = nil */), ...)
```

The command line flag is not connected to `config` and unusable because `Set` always returns an error.

I think you should prototype this in some real command and then we can discuss what of that prototype is worth moving to component-base/cli.

**Comment by aojea** [comment]
> The command line flag is not connected to `config` and unusable because `Set` always returns an error.

It does not return an error, it only checks for nil on the `NewMetaDuration` var to raise the error, the internal var with the metav1.Duration can be nil and is set without errors, there is a test case that covers that

> 
> I think you should prototype this in some real command and then we can discuss what of that prototype is worth moving to component-base/cli.

yes, is not working as expected, you are absolutely right about it, I have to do a proper test on a real component

/hold

**Comment by aojea** [comment]
doh, 

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1c0e26a]

goroutine 1 [running]:
k8s.io/kubernetes/cmd/kube-proxy/app.(*ProxyServer).setupConntrack(0xc000641f40)
	cmd/kube-proxy/app/server_others.go:342 +0x28a
k8s.io/kubernetes/cmd/kube-proxy/app.(*ProxyServer).platformSetup(0xc000641f40?)
	cmd/kube-proxy/app/server_others.go:94 +0x253
k8s.io/kubernetes/cmd/kube-proxy/app.newProxyServer(0xc000170840, {0x7fffc270e83d, 0x14})
	cmd/kube-proxy/app/server.go:642 +0x60b
k8s.io/kubernetes/cmd/kube-proxy/app.(*Options).Run(0xc000519b00)
	cmd/kube-proxy/app/server.go:390 +0x7f
k8s.io/kubernetes/cmd/kube-proxy/app.NewProxyCommand.func1(0xc0004fa700?, {0xc000641ea0?, 0x4?, 0x21cc31c?})
	cmd/kube-proxy/app/server.go:557 +0x285
github.com/spf13/cobra.(*Command).execute(0xc000004600, {0xc00013c010, 0xa, 0xa})
	vendor/github.com/spf13/cobra/command.go:940 +0x87c
github.com/spf13/cobra.(*Command).ExecuteC(0xc000004600)
	vendor/github.com/spf13/cobra/command.go:1068 +0x3a5
github.com/spf13/cobra.(*Command).Execute(...)
	vendor/github.com/spf13/cobra/command.go:992
k8s.io/component-base/cli.run(0xc000004600)
	vendor/k8s.io/component-base/cli/run.go:146 +0x2c5
k8s.io/component-base/cli.Run(0xc0000061a0?)
	vendor/k8s.io/component-base/cli/run.go:46 +0x17
main.main()
	cmd/kube-proxy/proxy.go:31 +0x18
I0908 09:45:19.072464       1 server_others.go:69] "Using iptables proxy"
I0908 09:45:19.072727       1 flags.go:64] FLAG: --bind-address="0.0.0.0"
I0908 09:45:19.072758       1 flags.go:64] FLAG: --bind-address-hard-fail="false"
I0908 09:45:19.072767       1 flags.go:64] FLAG: --boot-id-file="/proc/sys/kernel/random/boot_id"
I0908 09:45:19.072777       1 flags.go:64] FLAG: --cleanup="false"
I0908 09:45:19.072791       1 flags.go:64] FLAG: --cluster-cidr="10.64.0.0/14"
I0908 09:45:19.072809       1 flags.go:64] FLAG: --config=""
I0908 09:45:19.072816       1 flags.go:64] FLAG: --config-sync-period="15m0s"
I0908 09:45:19.072825       1 flags.go:64] FLAG: --conntrack-max-per-core="32768"
I0908 09:45:19.072833       1 flags.go:64] FLAG: --conntrack-min="131072"
I0908 09:45:19.072845       1 flags.go:64] FLAG: --conntrack-tcp-be-liberal="false"
I0908 09:45:19.072856       1 flags.go:64] FLAG: --conntrack-tcp-timeout-close-wait="1h0m0s"
I0908 09:45:19.072864       1 flags.go:64] FLAG: --conntrack-tcp-timeout-established="24h0m0s"
I0908 09:45:19.072870       1 flags.go:64] FLAG: --conntrack-udp-timeout=""
```

the udp flags is using an empty string, I'm super confused with how flags operate 😄

**Comment by aojea** [comment]
this was a fun exercise and one of the reasons to understand better all the flags vs config problematic, it seems flags is more sustainable and better for kubernetes deployments, since they are used directly on the manifest, instead of mounting the config from a configmap volume

**Pull request closed without merging**